File: Engine/Android/core/src/test/kotlin/despia/engine/FontsConformanceTest.kt
package despia.engine

import java.io.File
import org.junit.jupiter.api.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

/**
 * The font face-selection conformance runner - executes
 * OpenSource/Conformance/fonts/matching.json through THIS runtime's StackFonts
 * (parity/F01-fonts.md). The TS twin (@despia-native/kernel fonts.ts) and the Swift twin (StackFonts)
 * run the SAME file, so a type ramp cannot come out semibold on one renderer and bold on another.
 *
 * Missing corpus = loud failure - a silently-skipped conformance suite is how drift starts.
 */
class FontsConformanceTest {

    private fun corpusFile(): File {
        var dir = File(System.getProperty("user.dir")).absoluteFile
        while (true) {
            val candidate = File(dir, "OpenSource/Conformance/fonts/matching.json")
            if (candidate.isFile) return candidate
            dir = dir.parentFile ?: error("OpenSource/Conformance/fonts/matching.json not found")
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun root(): Map<String, Any?> {
        val root = json(corpusFile().readText()).foundationValue as? Map<String, Any?>
            ?: error("matching.json: not a JSON object")
        assertEquals(1, (root["version"] as? Number)?.toInt(), "matching.json version")
        return root
    }

    @Suppress("UNCHECKED_CAST")
    private fun section(name: String): List<Map<String, Any?>> =
        (root()[name] as? List<Map<String, Any?>>)?.also {
            assertTrue(it.isNotEmpty(), "$name must not be empty")
        } ?: error("matching.json: no $name[]")

    @Test
    @Suppress("UNCHECKED_CAST")
    fun weightMatchingAgreesWithCorpus() {
        for (case in section("matching")) {
            val name = case["name"] as? String ?: "<unnamed>"
            val faces = (case["faces"] as List<Any?>).map { (it as Number).toInt() }
            val request = (case["request"] as Number).toInt()
            val expected = (case["expect"] as? Number)?.toInt()
            assertEquals(expected, StackFonts.matchWeight(faces, request), name)
        }
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun slantSelectionAgreesWithCorpus() {
        for (case in section("italic")) {
            val name = case["name"] as? String ?: "<unnamed>"
            val faces = (case["faces"] as List<Map<String, Any?>>).map {
                StackFonts.Face((it["w"] as Number).toInt(), it["i"] as Boolean)
            }
            val request = case["request"] as Map<String, Any?>
            val got = StackFonts.selectFace(faces, (request["w"] as Number).toInt(), request["i"] as Boolean)
            val expect = case["expect"] as? Map<String, Any?>
            if (expect == null) {
                assertNull(got, name)
                continue
            }
            assertNotNull(got, "$name: expected a selection")
            assertEquals((expect["w"] as Number).toInt(), got.face.weight, "$name: weight")
            assertEquals(expect["i"] as Boolean, got.face.italic, "$name: italic")
            assertEquals(expect["synthesized"] as Boolean, got.synthesized, "$name: synthesized")
        }
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun variationClampingAgreesWithCorpus() {
        for (case in section("variation")) {
            val name = case["name"] as? String ?: "<unnamed>"
            val declared = (case["axes"] as? Map<String, Any?>)?.mapValues {
                val range = it.value as List<Any?>
                Pair((range[0] as Number).toDouble(), (range[1] as Number).toDouble())
            }
            val requested = (case["request"] as Map<String, Any?>).mapValues { (it.value as Number).toDouble() }
            val got = StackFonts.resolveVariation(declared, requested)
            val expect = case["expect"] as Map<String, Any?>
            val expectedApplied = (expect["applied"] as Map<String, Any?>)
                .mapValues { (it.value as Number).toDouble() }
            assertEquals(expectedApplied, got.applied, "$name: applied")
            assertEquals((expect["clamped"] as? List<Any?>)?.map { it as String } ?: emptyList(),
                         got.clamped, "$name: clamped")
            assertEquals((expect["dropped"] as? List<Any?>)?.map { it as String } ?: emptyList(),
                         got.dropped, "$name: dropped")
        }
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun variationParserAgreesWithCorpus() {
        for (case in section("parse")) {
            val name = case["name"] as? String ?: "<unnamed>"
            val expect = (case["expect"] as Map<String, Any?>).mapValues { (it.value as Number).toDouble() }
            assertEquals(expect, StackFonts.parseVariation(case["input"] as? String), name)
        }
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun featureParserAgreesWithCorpus() {
        for (case in section("features")) {
            val name = case["name"] as? String ?: "<unnamed>"
            val expect = (case["expect"] as List<Any?>).map { it as String }
            assertEquals(expect, StackFonts.parseFeatures(case["input"] as? String), name)
        }
    }
}

File: Web/packages/kernel/test/fonts-conformance.test.ts
//
//  fonts-conformance.test.ts — the SHARED font face-selection corpus through the TS kernel
//  (OpenSource/Conformance/fonts/matching.json). The Kotlin twin (:core FontsConformanceTest)
//  and the Swift twin (StackFonts) run the SAME file, so a type ramp cannot come out semibold on
//  one renderer and bold on another.
//
//  Missing corpus = loud failure — a silently-skipped conformance suite is how drift starts.
//

import { test } from "node:test";
import assert from "node:assert/strict";
import { readFileSync, existsSync } from "node:fs";
import { join, dirname, resolve } from "node:path";

import {
  matchFontWeight, selectFontFace, parseFontVariation, resolveFontVariation, parseFontFeatures,
  type FontFace,
} from "../src/index.ts";

function corpus(): { [k: string]: unknown } {
  let dir = resolve(import.meta.dirname ?? ".");
  for (;;) {
    const candidate = join(dir, "OpenSource/Conformance/fonts/matching.json");
    if (existsSync(candidate)) {
      const doc = JSON.parse(readFileSync(candidate, "utf-8")) as { [k: string]: unknown };
      assert.equal(doc["version"], 1, "matching.json: version");
      return doc;
    }
    const parent = dirname(dir);
    if (parent === dir) throw new Error("OpenSource/Conformance/fonts/matching.json not found");
    dir = parent;
  }
}

test("fonts: the CSS weight-matching algorithm agrees with the corpus", () => {
  const cases = corpus()["matching"] as { name: string; faces: number[]; request: number; expect: number | null }[];
  assert.ok(cases.length > 0);
  for (const c of cases) {
    assert.equal(matchFontWeight(c.faces, c.request), c.expect, c.name);
  }
});

test("fonts: slant selection and italic synthesis agree with the corpus", () => {
  type Case = {
    name: string;
    faces: { w: number; i: boolean }[];
    request: { w: number; i: boolean };
    expect: { w: number; i: boolean; synthesized: boolean } | null;
  };
  const cases = corpus()["italic"] as Case[];
  assert.ok(cases.length > 0);
  for (const c of cases) {
    const faces: FontFace[] = c.faces.map((f) => ({ weight: f.w, italic: f.i }));
    const got = selectFontFace(faces, c.request.w, c.request.i);
    if (c.expect === null) { assert.equal(got, null, c.name); continue; }
    assert.ok(got, `${c.name}: expected a selection`);
    assert.equal(got.face.weight, c.expect.w, `${c.name}: weight`);
    assert.equal(got.face.italic, c.expect.i, `${c.name}: italic`);
    assert.equal(got.synthesized, c.expect.synthesized, `${c.name}: synthesized`);
  }
});

test("fonts: variable-axis clamping agrees with the corpus", () => {
  type Case = {
    name: string;
    axes: { [k: string]: [number, number] } | null;
    request: { [k: string]: number };
    expect: { applied: { [k: string]: number }; clamped?: string[]; dropped?: string[] };
  };
  const cases = corpus()["variation"] as Case[];
  assert.ok(cases.length > 0);
  for (const c of cases) {
    const got = resolveFontVariation(c.axes, c.request);
    assert.deepEqual(got.applied, c.expect.applied, `${c.name}: applied`);
    assert.deepEqual([...got.clamped], c.expect.clamped ?? [], `${c.name}: clamped`);
    assert.deepEqual([...got.dropped], c.expect.dropped ?? [], `${c.name}: dropped`);
  }
});

test("fonts: the fontVariation parser agrees with the corpus", () => {
  const cases = corpus()["parse"] as { name: string; input: string; expect: { [k: string]: number } }[];
  assert.ok(cases.length > 0);
  for (const c of cases) {
    assert.deepEqual(parseFontVariation(c.input), c.expect, c.name);
  }
});

test("fonts: the fontFeature parser agrees with the corpus", () => {
  const cases = corpus()["features"] as { name: string; input: string; expect: string[] }[];
  assert.ok(cases.length > 0);
  for (const c of cases) {
    assert.deepEqual(parseFontFeatures(c.input), c.expect, c.name);
  }
});
